### PR TITLE
fix(identify): only report observed address once per connection

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

At the moment, `libp2p-identify` reports an observed address repeatedly as a **new** external address candidate even if it is the same address _from the same connection_. Unless the underlying transport supports roaming, a connection does not change its observed address.

We change the behaviour of `libp2p-identify` to remember the observed address for a particular connection and not re-emit the `NewExternalAddrCandidate` event for it. This allows users to probabilistically promote a candidate to an external address based on its report frequency. If an address is reported twice, it means we have two connections where the remote observed this address. Chances are, we have port-reuse enabled for this connection and it might thus be dialable or at least a good candidate for hole-punching.

Related: #4688.

## Notes & open questions

cc @altonen

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/identify/src/behaviour.rs [hunk 1]
```diff
@@ -48,6 +49,10 @@ pub struct Behaviour {
     config: Config,
     /// For each peer we're connected to, the observed address to send back to it.
     connected: HashMap<PeerId, HashMap<ConnectionId, Multiaddr>>,
+
+    /// the address a remote observed for us.
```
Comment (thomaseizinger) [inline protocols/identify/src/behaviour.rs]:
```suggestion
    /// The address a remote observed for us.
```

File: protocols/identify/src/behaviour.rs [hunk 2]
```diff
@@ -294,8 +300,27 @@ impl NetworkBehaviour for Behaviour {
...
+                    Entry::Occupied(mut already_observed) => {
+                        log::info!(
+                            "Our observed address on connection {id} changed from {} to {observed}",
+                            already_observed.get()
+                        );
+
+                        *already_observed.get_mut() = observed.clone();
+                        self.events
+                            .push_back(ToSwarm::NewExternalAddrCandidate(observed));
+                    }
```
Comment (thomaseizinger) [inline protocols/identify/src/behaviour.rs:297]: This is the "roaming connection" usecase. Not sure if it is worth supporting?

**Comment by thomaseizinger** [comment]
Applying `internal-change` because we just adjust the tests.

File: misc/metrics/CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,7 @@
 ## 0.14.0 - unreleased
 
+- Add metrics for `SwarmEvent::{NewExternalAddrCandidate,ExternalAddrConfirmed,ExternalAddrExpired}`.
+  See [PR 4721](https://github.com/libp2p/rust-libp2p/pull/4721).
```
Comment (mxinden) [inline misc/metrics/CHANGELOG.md:4]: :pray:

File: protocols/identify/CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,10 @@
 
 - Add `Info` to the `libp2p-identify::Event::Pushed` to report pushed info.
   See [PR 4527](https://github.com/libp2p/rust-libp2p/pull/4527)
+- Don't repeatedly report the same observed address as a `NewExternalAddrCandidate`.
+  Instead, only report each observed address once.
```
Comment (mxinden) [inline protocols/identify/CHANGELOG.md]:
```suggestion
  Instead, only report each observed address once per connection.
```

Would be more correct, no?

File: protocols/identify/src/behaviour.rs (hunk 2 above)
Comment (mxinden) [reply protocols/identify/src/behaviour.rs:297]: I don't think it adds much noise. In other words, I am fine keeping it here.

File: protocols/identify/tests/smoke.rs [hunk 5]
```diff
@@ -77,6 +80,75 @@ async fn periodic_identify() {
         other => panic!("Unexpected events: {other:?}"),
     }
 }
+#[async_std::test]
+async fn only_emits_address_candidate_once_per_connection() {
```
Comment (mxinden) [inline protocols/identify/tests/smoke.rs:85]: :+1:

**Review by mxinden — APPROVED** [review]
Good to go from my end. Thanks.

**Comment by thomaseizinger** [comment]
@mxinden I updated some docs as well.

**Review by mxinden — APPROVED** [review]
Docs read well to me.

**Comment by mxinden** [comment]
Retriggered CI with "spurious network failure".

**Pull request merged**